### PR TITLE
Refactor commit lock mechanism from HiveTableOperations

### PR DESCRIPTION
**Pull request by szlta**
`HiveTableOperations` has grown to a substantial size and got pretty complex.
A large portion of the logic is the implementation of the locking mechanism which I think could be separated out to reduce this complexity.

Let me know what you think @pvary @szehon-ho

### CODE REVIEW

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCommitLock.java [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    long lockId = lockResponse.getLockid();
+    this.hmsLockId = Optional.of(lockId);
+
+    final long start = System.currentTimeMillis();
+    long duration = 0;
+    boolean timeout = false;
+
+    try {
+      if (state.get().equals(LockState.WAITING)) {
+        // Retry count is the typical "upper bound of retries" for Tasks.run() function. In fact,
```
Comment (pvary) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCommitLock.java:170]: Nit: could you please reformat the comment?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCommitLock.java [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+import org.apache.iceberg.ClientPool;
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.iceberg.util.Tasks;
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class HiveCommitLock {
```
Comment (pvary) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCommitLock.java:49]: Would it be useful to implement the standard `java.util.concurrent.locks.Lock` interface?

**Comment by pvary** [comment]
@szlta: any particular change to highlight for the review? Also, please check the failing tests

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCommitLock.java [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+                    LockResponse response = metaClients.run(client -> client.checkLock(id));
+                    LockState newState = response.getState();
+                    state.set(newState);
+                    if (newState.equals(LockState.WAITING)) {
+                      throw new WaitingForLockException(
+                          String.format(
+                              "Waiting for lock on table %s.%s", databaseName, tableName));
+                    }
+                  } catch (InterruptedException e) {
+                    Thread.interrupted(); // Clear the interrupt status flag
```
Comment (szehon-ho) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCommitLock.java:197]: Again I see its copy and paste, but its weird, it seems this is checking a boolean and ignoring it.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCommitLock.java [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+          "Could not acquire the lock on %s.%s, " + "lock request ended in state %s",
+          databaseName, tableName, state);
+    }
+
+    hiveLockHeartbeat = new HiveLockHeartbeat(metaClients, lockId, lockHeartbeatIntervalTime);
+    hiveLockHeartbeat.schedule(exitingScheduledExecutorService);
+  }
+
+  private void releaseHmsLock() {
+    if (hmsLockId.isPresent()) {
```
Comment (szehon-ho) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCommitLock.java:233]: If we have a simple boolean acquired (as mentioned in my other comment), we can just replace these theese two null checks with a single check on release().  As these are both private methods, checking both seems excessive?

**Review by szehon-ho — COMMENTED** [review]
I think its good to refactor it, but definitely need to spend a bit of time to review this critical code.  Gave a pass

**Pull request closed without merging**